gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs

```diff
@@ -0,0 +1,704 @@
+        .state
+        .update_raw(key, value)
+        .expect("gw_store");
+    SUCCESS
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn gw_load(key_addr: *const u8, value_addr: *mut u8) -> c_int {
+    let key = load_data_h256(key_addr);
+    let val = HOST.lock().unwrap().state.get_raw(&key).expect("gw_load");
+    store_h256(value_addr, &val);
+    SUCCESS
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn gw_set_return_data(addr: *const u8, len: u64) -> c_int {
+    let buf = load_bytes(addr, len);
+    HOST.lock().unwrap().run_result.return_data = buf;
+    SUCCESS
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn gw_create(
+    script_addr: *const u8,
+    script_len: u64,
+    account_id_addr: *mut u32,
+) -> c_int {
+    let script_data = load_bytes(script_addr, script_len);
+    let script = match Script::from_slice(&script_data[..]) {
+        Ok(script) => script,
+        Err(_err) => return GW_ERROR_INVALID_ACCOUNT_SCRIPT.into(),
+    };
+    let script_hash = script.hash();
+    let mut host = HOST.lock().unwrap();
+    // check exists
+    match host.get_account_id_by_script_hash(&script_hash.into()) {
+        Ok(Some(_)) => return GW_ERROR_DUPLICATED_SCRIPT_HASH.into(),
+        Ok(None) => {}
+        Err(_err) => {
+            return ERROR;
+        }
+    }
+    // TODO: valide script
+    let id = match host.get_account_count() {
+        Ok(id) => id,
+        Err(_err) => {
+            return ERROR;
+        }
+    };
+
+    let account_nonce_key = build_account_field_key(id, GW_ACCOUNT_NONCE_TYPE);
+    host.state
+        .update_raw(account_nonce_key, H256::zero())
+        .expect("account nonce key");
+    let account_script_hash_key = build_account_field_key(id, GW_ACCOUNT_SCRIPT_HASH_TYPE);
+    host.state
+        .update_raw(account_script_hash_key, script_hash)
+        .expect("account script hash key");
+    // script hash to id
+    let script_hash_to_id_value: H256 = {
+        let mut buf: [u8; 32] = H256::from_u32(id).into();
+        // the first 4 bytes is id, set exists flag(fifth byte) to 1
+        buf[4] = 1;
+        buf.into()
+    };
+    let script_hash_to_account_id_key = build_script_hash_to_account_id_key(&script_hash[..]);
+    host.state
+        .update_raw(script_hash_to_account_id_key, script_hash_to_id_value)
+        .expect("write script hash to account id key");
+    // insert script
+    host.state.insert_script(script_hash, script);
+    host.state
+        .set_account_count(id + 1)
+        .expect("set account count");
+
+    let size_ptr = account_id_addr.as_mut().expect("casting pointer");
+    *size_ptr = id;
+
+    SUCCESS
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn gw_load_tx(addr: *mut c_void, len: *mut u64) -> c_int {
+    if let Some(tx) = &HOST.lock().unwrap().tx {
+        store_data(addr, len, 0, tx);
+        SUCCESS
+    } else {
+        GW_ITEM_MISSING
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn gw_load_block_info(addr: *mut c_void, len: *mut u64) -> c_int {
```

Comment:
Some extern functions do not require unsafe.

> https://doc.rust-lang.org/book/ch19-01-unsafe-rust.html?highlight=no_mangle#calling-rust-functions-from-other-languages

Reply:
Being `extern` does not require the functions to be `unsafe`, but many of these functions are actually `unsafe`, i.e. it's possible to cause UB if they are called by safe rust functions with incorrect parameters.